### PR TITLE
feat: Rm workspace crates version before caching

### LINKED ISSUE #146: Rm workspace from Cargo.lock before caching
In [cargo-binstall](https://github.com/cargo-bins/cargo-binstall), we sometimes add dependency which is already present in Cargo.lock to one of the workspace.

We also have a release workflow where we create one PR for each workspace.

In both cases, Swatinem/rust-cache would upload a new cache for each pull request while nothing has really changed, which really overblown our GitHub Cache as we have >10 jobs and each cache artifact is roughly ~200MB large after compression.

I believe it can be fixed by creating a copy of Cargo.lock and remove all workspace crates from it.

Or can go one step further, create Swatinem/rust-cache's own lockfile which only contains the dependencies and their version, since this action don't care about the dependency graph, rather only the dependencies and their versions

### PR DESCRIPTION
**Pull request by NobodyXu**
Fixed #146

 - Set all `package.version` in `Cargo.toml` to `0.0.0`
 - Set `{build-, dev-, }dependencies` of workspace crates to `0.0.0`
 - Remove workspace crates from `Cargo.lock` before caching
 - Sort all toml objects before hashing them as json

Signed-off-by: Jiahao XU <Jiahao_XU@outlook.com>

### CODE REVIEW

**Review by Swatinem — COMMENTED** [review]
This is a really interesting approach, and I believe this works, as we are indeed only interested in non-workspace crates.

Can you check that the relevant files are still listed in the info output of the action?

Also, I think you would have to do the same filtering/nulling of workspace dependencies in the `Cargo.toml` files, as you can have dependencies such as `workspace-crate = { path = "path/for/local/development", version = "1.0.0-version-for-installing-via-cratesio" }`

**Comment by NobodyXu** [comment]
> Can you check that the relevant files are still listed in the info output of the action?

Sorry I am confused.
Can you elaborate on what you would like me to check for?

**Comment by NobodyXu** [comment]
> Also, I think you would have to do the same filtering/nulling of workspace dependencies in the `Cargo.toml` files, as you can have dependencies such as `workspace-crate = { path = "path/for/local/development", version = "1.0.0-version-for-installing-via-cratesio" }`

Thanks!
I've updated the PR and also sort the objects before using `JSON.stringify` to convert them to strings for hashing.

File: src/config.ts [hunk 1]
```diff
@@ -128,23 +130,93 @@ export class CacheConfig {
...
     let keyFiles = await globFiles("rust-toolchain\nrust-toolchain.toml");
+
+    hasher = crypto.createHash("sha1");
+
     for (const workspace of workspaces) {
       const root = workspace.root;
       keyFiles.push(
         ...(await globFiles(
-          `${root}/**/Cargo.toml\n${root}/**/Cargo.lock\n${root}/**/rust-toolchain\n${root}/**/rust-toolchain.toml`,
+          `${root}/**/rust-toolchain\n${root}/**/rust-toolchain.toml`,
```
Comment (NobodyXu) [inline src/config.ts:140]: P.S. I just realized that `rust-cache` does not include `.cargo/config.toml` in hashing, I will submit a PR to fix this after this PR is merged.

Comment (NobodyXu) [reply src/config.ts:140]: [`.cargo/config.toml`](https://doc.rust-lang.org/cargo/reference/config.html#configuration-format) contains some settings that can affect the build, like `patch`, `profile`, `build.rustc-wrapper`, `env`, `registeries`, `target.<triple>` and etc.

Comment (Swatinem) [reply src/config.ts:140]: ah, that is a very good idea indeed!

**Comment by Swatinem** [comment]
> Sorry I am confused. Can you elaborate on what you would like me to check for?

By that I mean the summary information that is printed by the action, this thing here:

```

Cache Configuration
  Workspaces:
      /home/runner/work/rust-cache/rust-cache/tests
  Cache Paths:
      /home/runner/.cargo
      /home/runner/work/rust-cache/rust-cache/tests/target
  Restore Key:
      v0-rust-coverage-0308f3c6
  Cache Key:
      v0-rust-coverage-0308f3c6-f80d3fb4
  .. Prefix:
    - v0-rust-coverage
  .. Environment considered:
    - Rust Version: 1.69.0 x86_64-unknown-linux-gnu (84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc)
    - CARGO_INCREMENTAL
    - CARGO_TERM_COLOR
  .. Lockfiles considered:
    - /home/runner/work/rust-cache/rust-cache/tests/Cargo.lock
    - /home/runner/work/rust-cache/rust-cache/tests/Cargo.toml
    - /home/runner/work/rust-cache/rust-cache/tests/wasm-workspace/Cargo.lock
    - /home/runner/work/rust-cache/rust-cache/tests/wasm-workspace/Cargo.toml
    - /home/runner/work/rust-cache/rust-cache/tests/wasm-workspace/crates/one/Cargo.toml
    - /home/runner/work/rust-cache/rust-cache/tests/wasm-workspace/crates/two/Cargo.toml
```

It should still list these files in the "lockfiles considered" section. Even though you are not hashing the literal file, but rather using more targeted information from them.

File: src/config.ts [hunk 2]
```diff
@@ -272,3 +344,20 @@ async function globFiles(pattern: string): Promise<string[]> {
   });
   return await globber.glob();
 }
+
+function sort_object(o: any): any {
```
Comment (Swatinem) [inline src/config.ts:348]: hm, tbh I wouldn’t go this far, but it doesn’t hurt either.

File: src/config.ts [hunk 3]
```diff
@@ -128,23 +130,93 @@ export class CacheConfig {
...
       keyFiles.push(
         ...(await globFiles(
-          `${root}/**/Cargo.toml\n${root}/**/Cargo.lock\n${root}/**/rust-toolchain\n${root}/**/rust-toolchain.toml`,
+          `${root}/**/rust-toolchain\n${root}/**/rust-toolchain.toml`,
         )),
       );
+
+      // Get name of workspace crates.
+      const ws_crates = new Set(
+        (await workspace.getPackages()).map(p => p["name"])
```
Comment (Swatinem) [inline src/config.ts:146]: This here is a problem, as it internally calls `cargo metadata`.

We had to remove a call that happens in the "restore"/pre-phase of the action:
https://github.com/Swatinem/rust-cache/commit/3f2513fdf484da3bcfe60b1b40921055681bcdbf#diff-32b32ae92adc890fbd79b1d74f84cf0351380d3e0282f9a9f01352763140541dL46

This was because under some circumstances, probably because of a bug in cargo itself, `cargo metadata` will actually go ahead and compile a bunch of stuff. And by "compile a bunch of stuff", it did a full `cargo build`. Don’t ask me why because it is a mystery to me, but we experienced that problem on the https://github.com/getsentry/relay repo.

Long story short: `getPackages` is a big no-no in the pre-phase. (might be good to even add a big comment that function explaining this)

This is a bit unfortunate, but I believe the workaround for that is very simple:
Just check if you have a `path` dependency, and zero-out the version of that. The way cargo works, `path` dependencies are only a thing in local workspaces and always preferred over versions. Once you publish and someone else installs the package as a dependency, the `version` will be used instead.

Comment (Swatinem) [reply src/config.ts:146]: > Just check if you have a `path` dependency, and zero-out the version of that.

well, that wouldn’t work for the lockfile though :-(

Comment (Swatinem) [reply src/config.ts:146]: Actually… We are iterating all the `Cargo.toml` files, so we don’t need the `cargo metadata` invocation in that case because iterating over the `toml` files already gives us all the workspace dependency names. And we don’t have to care about all the target names either.

File: src/config.ts (hunk 2 above)
Comment (NobodyXu) [reply src/config.ts:348]: I just want to make sure everything is normalised to a fix order, so if the users only change the order it should not save a new cache.

File: src/config.ts (hunk 3 above)
Comment (NobodyXu) [reply src/config.ts:146]: Thanks!
And I just realized that you can rename dependencies in `Cargo.toml`:

```
serde2 = { package = "serde", version = "1" }
serde = { path = "...", version = "1" }
```

Looking at the `Cargo.lock`, I found that dependencies with `path` does not have `[[package]].source`, so we can figure that out all by looking at `Cargo.lock` or `Cargo.toml` independently!

Comment (NobodyXu) [reply src/config.ts:146]: Also found that `[[package]].checksum` will not present for workspace dep, so I will detect them by looking at both `source` and `checksum`.

Comment (NobodyXu) [reply src/config.ts]: Thanks, I've pushed the fix.

**Comment by NobodyXu** [comment]
> It should still list these files in the "lockfiles considered" section. Even though you are not hashing the literal file, but rather using more targeted information from them.

Thanks!
I've fixed this by adding these files to `self.keyFiles` after hashing is done.

**Review by Swatinem — APPROVED** [review]

**Comment by NobodyXu** [comment]
@Swatinem Sorry I forgot to update the `dist/` using `npm run prepare`, I've updated it now and it should be ready to merge.

**Comment by Swatinem** [comment]
Thanks for sticking with this, great work!

File: src/config.ts (hunk 1 above)
Comment (NobodyXu) [reply src/config.ts:141]: I've opened #149 for this.
